File: ruby/titulo_propiedad.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

equiere_relative "jugador"

module Civitas

  class TituloPropiedad
    
    attr_accessor :nombre, :alquiler_base, :factor_revalorizacion, :hipoteca_base, :precio_compra, :precio_edificar, :hipotecado, :num_casas, :num_hoteles, :propietario
    
    def initialize(nombre, alquiler_base, factor_revalorizacion, hipoteca_base, precio_compra, precio_edificar)
      @nombre = nombre
      @alquiler_base
      @factor_revalorizacion = factor_revalorizacion
      @hipoteca_base = hipoteca_base
      @precio_compra = precio_compra
      @precio_edificar = precio_edificar
      @hipotecado = false
      @num_casas = 0
      @num_hoteles = 0
      @propietario = nil
      @@factor_interes_hipoteca = 1.1
    end
    
    def to_string()
      cad = "La propiedad #{@nombre}, tiene un precio de alquiler base de #{@alquiler_base}, un factor de revalorizacion #{@factor_revalorizacion}, un precio de hipoteca de #{@hipoteca_base}, un precio de compra de #{@precio_compra}, un precio de edificación de #{@precio_edificar}"
      return cad
    end
    
    def get_precio_alquiler()
      if (@hipotecado || propietario_encarcelado())
        precio = 0
      else
        precio = alquiler_base*(1+(@num_casas*0.5)+(num_hoteles*2.5)) 
      end
            
      return precio
    end
    
    def get_importe_cancelar_hipoteca()
      return @hipoteca_base*@@factor_intereses_hipoteca
    end
    
    def tramitar_alquiler(jugador)
      if (@propietario != nil && @propietario != jugador)
        precio = get_precio_alquiler()
        
        #Implementar métodos
        jugador.paga_alquiler(precio)
        @propietario.recibe(precio)
      end
    end
    
    def propietario_encarcelado()
      #Implementar clase Jugador y métodos
      return @propietario.encarcelado()
    end
    
    def cantidad_casas_hoteles()
      return @num_casas+@num_hoteles
    end
    
    def get_precio_venta()
      return (@precio_compra + @precio_edificar*(@num_casas+@num_hoteles)*@factor_revalorizacion)
    end
    
    def derruir_casas(n, jugador)
      hecho = false
      es_propietario = @propietario == jugador
      if (es_propietario && @num_casas > n)
        @num_casas -= n
        hecho = true        
      end
      
      return hecho
    end
    
    def vender(jugador)
      
      hecho = false
      
      if (@propietario == jugador && !@hipotecado)
        jugador.recibe(get_precio_venta())
        @propietario = nill
        @num_hoteles = 0
        @num_casas = 0
        hecho = true
      end
      
      return hecho
    end
    
    def es_este_el_propietario(jugador)
      return @propietario == jugador
    end
    
    def tiene_propietario()
      return (@propietario != nill)
    end
    
  end
end

File: java/civitas/TituloPropiedad.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package civitas;

/**
 *
 * @author mota
 */
public class TituloPropiedad {
    
    private String nombre;
    private float precioBaseAlquiler;
    private float factorRevalorizacion;
    private float precioBaseHipoteca;
    private float precioCompra;
    private float precioEdificar;
    
    private float factorInteresesHipoteca = 1.1f; //la f del final indica que es un float
    private boolean hipotecado;
    private int numCasas;
    private int numHoteles;
    
    //asociacion de la clase jugador
    Jugador propietario;
    
    //constructor
    TituloPropiedad(String nom, float precioBaseA, float factorR, float precioBaseH , float precioC, float precioE){
        nombre = nom;
        precioBaseAlquiler = precioBaseA;
        factorRevalorizacion = factorR;
        precioBaseHipoteca = precioBaseH;
        precioCompra = precioC;
        precioEdificar = precioE;
        
        //inicializados a 0 y a false 
        hipotecado = false;
        numCasas = 0;
        numHoteles = 0;
        
        //asociacion de la clase jugador a null
        propietario = null;
    }
    
    void actualizarPropietarioPorConversion(Jugador jugador){
        propietario=jugador;
    }
    
    
    boolean cancelarHipoteca(Jugador jugador){
        boolean resultado = false;
        
        if (getHipotecado()){
            if(esEsteElPropietario(jugador)){
                propietario.paga(getImporteCancelarHipoteca());
                hipotecado = false;
                resultado = true;
            }
        }
        
        return resultado;
    }

    
    int cantidadCasasHoteles(){
        return numCasas + numHoteles;
    }
    
    
    boolean comprar(Jugador jugador){
        boolean resultado = false;
        if(!tienePropietario()){
            propietario = jugador;
            propietario.paga(precioCompra);
            resultado = true;
        }
        
        return resultado;
    }
    
    
    boolean construirCasa(Jugador jugador){
        boolean resultado = false;
        
        if (esEsteElPropietario(jugador)){
            propietario.paga(precioEdificar);
            numCasas++;
            resultado = true;
        }
        
        return resultado;
    }
    
    
    
    boolean construirHotel(Jugador jugador){
        boolean resultado = false;
        
        if (esEsteElPropietario(jugador)){
            propietario.paga(precioEdificar);
            numHoteles++;
            resultado = true;
        }
        
        return resultado;
    }
    
    
    boolean derruirCasas(int n, Jugador jugador){
        //booleano para saber si se ha derruido la casa o no
        boolean derruido = false;
        
        if(jugador == propietario && numCasas >= n){
            numCasas -= n;
            derruido = true;
        }
        
        return derruido;
    }
    
    private boolean esEsteElPropietario(Jugador jugador ){
        boolean EsPropietario = false;
        
        if(jugador == propietario){
            EsPropietario = true;
        }
        
        return EsPropietario;
    }
    
    public boolean getHipotecado(){
         return hipotecado;
    }
    
    float getImporteCancelarHipoteca(){
        float precioHipoteca = precioBaseAlquiler*(1+(numCasas*0.5f)+(numHoteles*2.5f));
        return precioHipoteca * factorInteresesHipoteca;
    }
    
    private float getImporteHipoteca(){
        return precioBaseAlquiler*(1+(numCasas*0.5f)+(numHoteles*2.5f));
    }
    
    public String getNombre(){
        return nombre;
    }
    
    int getNumCasas(){
        return numCasas;
    }
    
    int getNumHoteles(){
        return numHoteles;
    }
    
    //esta función referencia a la clase jugador
    private boolean propietarioEncarcelado(){
        boolean encarcelado = true;
        
        if(!propietario.encarcelado){
            encarcelado = false;
        }
        
        return encarcelado;
    }
    
    private float getPrecioAlquiler(){
        float alquiler;
        
        if(!propietarioEncarcelado() && !hipotecado){
            alquiler = precioBaseAlquiler*(1+(numCasas*0.5f)+(numHoteles*2.5f));
        }
        else
            alquiler = 0;
        
        return alquiler;
    }
    
    float getPrecioCompra(){
        return precioCompra;
    }
    
    float getPrecioEdificar(){
        return precioEdificar;
    }
    
    private float getPrecioVenta(){
        return precioCompra+((cantidadCasasHoteles()*precioEdificar)*factorRevalorizacion);
    }
    
    Jugador getPropietario(){
        return propietario;
    }
    
    
    boolean hipotecar(Jugador jugador){
        boolean resultado = false;
        
        if (!propietario.isEncarcelado() && !hipotecado && esEsteElPropietario(jugador)){
            propietario.recibe(getImporteHipoteca());
            hipotecado = true;
            resultado = true;
        }
        
        return resultado;
    }
    
    
    boolean tienePropietario(){
        return propietario!=null;
    }
   
    @Override
    public String toString(){
        String parametros = "Nombre :"+nombre+" Alquiler base : "+precioBaseAlquiler+" precio base de la hipoteca "
                +precioBaseHipoteca+" factor de revalorización: "+factorRevalorizacion+" factor de la compra : "
                +precioCompra+" precio para edificar "+precioEdificar;
        return parametros;
    }
    
    
    void TramitarAlquiler(Jugador jugador){
        if(tienePropietario() && esEsteElPropietario(jugador)){
            float alquiler = getPrecioAlquiler();
            jugador.pagaAlquiler(alquiler);
            propietario.recibe(alquiler);
        }
    }
    
    boolean vender(Jugador jugador){
        boolean venta = false;
        
        if(esEsteElPropietario(jugador) && !this.hipotecado){
            jugador.recibe(getPrecioVenta());
            propietario=null;
            numHoteles=0;
            numCasas=0;
            venta=true;
        }
        
        return venta;
    }
}
